### DISCUSSION TITLE
What is a framework ?

### DISCUSSION THREAD

**Discussion by waghanza** [the-benchmarker/web-frameworks#3641] [Ideas]
A framework is a set of components working together. 

There is a huge variety of frameworks, and each of them have some pros/cons (and having them in mind is a requirement to analyze results produced here).

#### We can classify frameworks depending on the scope covered

+ Some frameworks have components to deal with data storage (usually a database), and then we can say that those are **full-stack**. For example, `rails` is the most widely spread **full-stack** framework in the ruby community.
+ Some frameworks deals only with routing (redirect URL to any code) and then are considered as **micro**. As an example, `flask` is considered as **micro** framework in `python`

#### We can classify framework based on its design and conventions

+ Some frameworks are considered as **opiniated**. This mean the authors makes some assumptions about architecture, best practices, and convention the developers should use. As an example, `spring` could be considered as an **opiniated** framework in the `java` community.
+ On the contrary, some frameworks are considered as **non opiniated**. It means that developers who use this frameworks have to make their own assumptions about architecture, designed and best practices. For example, `scotty` is considered as a **non opiniated** framework in `haskell`

#### We can classify frameworks based depending on the components in used

+ Some newly created framework are providing the whole development toolbelt. Those kind of frameworks let the developers choose some components, and run on top of them. `Quarkus` in `java` is a perfect example for that. We can not classify them yet (but do we realy need ?)

**Comment by System-Glitch** [comment]
A definition of framework in that case would be that it has more than one purpose. For example, gorilla/mux is not a framework because its only purpose is routing, therefore, it is a library.
Another definition which as given in the original issue was that a framework calls your code, while a library is called by your code. In other words, a framework directs how your code will be executed, a library doesn't.

My own definition differs but I don't think there is any real consensus over the exact meaning of "framework". In my opinion, a framework is necessarily opinionated, because of the etymology of the word "framework".

I like your definition initiative, because it is important to think about what the library/framework does in the background while checking the results. It is normal that a full-stack framework is slower than a micro library. Adding this kind of information or grouping each type of library/framework together in the results would make them more relevant. 👍🏻

**Reply by waghanza** [reply]
I understand the idea, but the most important is ... to be understand by non technical person

I mean a purpose of this definition is to be agnostic of your skills. I am not sure that anyone would understand
>  a framework calls your code, while a library is called by your code

**Reply by System-Glitch** [reply]
How would you qualify a non technical person, and why would they check benchmarks? It's usually tech people who pick which tech is used for their projects.

If you want to make the definition as simple as possible, then keeping it to framework (micro/full-stack) and library is probably best.

**Reply by waghanza** [reply]
> How would you qualify a non technical person

A person who is not directly involved with writing source code

> and why would they check benchmarks? 

It's probably not the benchmark part that is interesting, in this case, but the knowledge such tools exists (despite the benchmark, I want also this list to be a kind of promotion.

> If you want to make the definition as simple as possible, then keeping it to framework (micro/full-stack) and library is probably best.

I prefer micro framework over library. Since a micro framework use, at least two components : one for parsing http requests, and an other for parsing URL
Q

**Comment by ahopkins** [comment]
> We can classify frameworks based depending on the components in used

I think this is too much. This to me sounds more like a footnote, or something that could be part of a description. It is more akin to "opinionated"-ness. Not necessarily in the *how* to build something, but the *with what* to build with. Too many layers of classifications borders on the unusable.

I like the 2 x 2 grid potential with full v micro, and opinionated v non.

```
            | opinionated | non-opinionated
------------+-------------+-----------------
full-stack  |             |
------------+-------------+-----------------
micro       |             |
```

**Comment by waghanza** [comment]
Perhaps tags are more flexible @ahopkins 

The idea is to use a separate API in a mid-term future
